src/subcommands/clean_branches.py

```diff
@@ -0,0 +1,86 @@
+# coding=utf-8
+from __future__ import print_function, unicode_literals
+
+import os
+import sys
+from argparse import ArgumentParser, Namespace
+
+from git_helpers import get_branch_tracker, get_current_branch, git, does_branch_exist, BranchTracker
+from subcommands.base_command import BaseCommand
+
+try:
+    # noinspection PyUnresolvedReferences
+    from typing import (
+        Iterable,
+        List,
+        Tuple,
+        TypeVar,
+        Text,
+    )
+
+    T = TypeVar('T')
+except ImportError:
+    pass
+
+from git_helpers import get_branch_tracker, get_current_branch, BranchTracker
+
+class CleanBranches(BaseCommand):
+    def get_name(self):
+        # type: () -> Text
+        return 'clean-branches'
+
+    def get_short_description(self):
+        # type: () -> Text
+        return 'deletes branches that are no longer known to git'
+
+    def inflate_subcommand_parser(self, parser):
+        # type: (ArgumentParser) -> None
+        parser.add_argument("-a", "--archive", action="store_true",
+                            help="set to archive instead of delete invalid branches")
+
+    def run_command(self, args):
+        # type: (Namespace) -> None
+        clean_invalid_branches(args.archive)
+
+def clean_invalid_branches(archive):
+    # type: (bool) -> None
+    with get_branch_tracker() as tracker:
+        roots = []
+        for parent in tracker.get_all_parents():
+            if not tracker.has_parent(parent):
+                roots.append(parent)
+
+        roots = sorted(roots)
+
+        for root in roots:
+            clean_invalid_branches_internal(tracker, root, archive)
+
+def clean_invalid_branches_internal(tracker, branch_name, archive):
+    # type: (BranchTracker, Text, bool) -> None
+    for child in tracker.children_for_parent(branch_name):
+        clean_invalid_branches_internal(tracker, child, archive)
+
+    if is_branch_invalid(tracker, branch_name):
+        if archive:
+            archive_invalid_branch(tracker, branch_name)
+        else:
+            delete_invalid_branch_if_possible(tracker, branch_name)
+
+def is_branch_invalid(tracker, branch_name):
+    # type: (BranchTracker, Text) -> bool
+    return not does_branch_exist(branch_name) and not tracker.is_archived(branch_name)
+
+def archive_invalid_branch(tracker, branch_name):
+    # type: (BranchTracker, Text) -> None
+    print("Archiving invalid branch {}".format(branch_name))
+    tracker.set_is_archived(branch_name, True)
+
```

Comment:
nit: There should be 2 blank lines here instead of 1, according to PEP 8 (eg needs another line)

Reply:
Missed this one.